examples/marketplace-contract/src/cis2_client.rs

```diff
@@ -0,0 +1,185 @@
+        T: IsTokenId + Clone + Copy,
+        A: IsTokenAmount + Clone + Copy + ops::Sub<Output = A>,
+    >(
+        host: &mut impl HasHost<State<S, T, A>, StateApiType = S>,
+        nft_contract_address: &ContractAddress,
+    ) -> Result<bool, Cis2ClientError> {
+        let params = SupportsQueryParams {
+            queries: vec![StandardIdentifierOwned::new_unchecked("CIS-2".to_string())],
+        };
+        let parsed_res: SupportsQueryResponse = Cis2Client::invoke_contract_read_only(
+            host,
+            nft_contract_address,
+            SUPPORTS_ENTRYPOINT_NAME,
+            &params,
+        )?;
+        let supports_cis2: bool = {
+            let f = parsed_res
+                .results
+                .first()
+                .ok_or(Cis2ClientError::InvokeContractError)?;
+            match f {
+                SupportResult::NoSupport => false,
+                SupportResult::Support => true,
+                SupportResult::SupportBy(_) => false,
+            }
+        };
+
+        Ok(supports_cis2)
+    }
+
+    pub(crate) fn is_operator_of<
+        S: HasStateApi,
+        T: IsTokenId + Clone + Copy,
+        A: IsTokenAmount + Clone + Copy + ops::Sub<Output = A>,
+    >(
+        host: &mut impl HasHost<State<S, T, A>, StateApiType = S>,
+        owner: Address,
+        current_contract_address: ContractAddress,
+        nft_contract_address: &ContractAddress,
+    ) -> Result<bool, Cis2ClientError> {
+        let params = &OperatorOfQueryParams {
+            queries: vec![OperatorOfQuery {
+                owner,
+                address: Address::Contract(current_contract_address),
+            }],
+        };
+
+        let parsed_res: OperatorOfQueryResponse = Cis2Client::invoke_contract_read_only(
+            host,
+            nft_contract_address,
+            OPERATOR_OF_ENTRYPOINT_NAME,
+            params,
+        )?;
+
+        let is_operator = parsed_res
+            .0
+            .first()
+            .ok_or(Cis2ClientError::InvokeContractError)?
+            .to_owned();
+
+        Ok(is_operator)
+    }
+
+    pub(crate) fn get_balance<
+        S,
+        T: IsTokenId + Clone + Copy,
+        A: std::default::Default + IsTokenAmount + Clone + Copy + ops::Sub<Output = A>,
+    >(
+        host: &mut impl HasHost<State<S, T, A>, StateApiType = S>,
+        token_id: T,
+        nft_contract_address: &ContractAddress,
+        owner: Address,
+    ) -> Result<A, Cis2ClientError>
+    where
+        S: HasStateApi,
+    {
+        let params = BalanceOfQueryParams {
+            queries: vec![BalanceOfQuery {
+                token_id,
+                address: owner,
+            }],
+        };
+
+        let parsed_res: BalanceOfQueryResponse<A> = Cis2Client::invoke_contract_read_only(
+            host,
+            nft_contract_address,
+            BALANCE_OF_ENTRYPOINT_NAME,
+            &params,
+        )?;
+
+        let ret = parsed_res.0.first().map_or(A::default(), |f| *f);
```

Comment:
Default should never be needed here. The response must contain a single value, otherwise this should fail. Using `default` here is a potential for problems since it silently does the wrong thing.